File: card-shuffling/card-shuffling.py
import random
import math

def card_shuffle(n):
    total = 52

    probs = {}
    for i in range(0, total+1):
        prob = math.comb(total, i) * 0.5**total
        probs[i] = prob

    suits = ["H", "D", "C", "S"]
    ranks = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]

    deck = [f"{rank} {suit}" for suit in suits for rank in ranks]
    for _ in range(n):
        left = random.choices(list(probs.keys()), weights=list(probs.values()), k=1)[0]
        right = total - left
        leftArr = deck[:left]
        rightArr = deck[left:]

        newDeck = []
        for _ in range(total):
            num = random.randint(1, left + right)
            if num > left:
                newDeck.append(rightArr[0])
                rightArr.pop(0)
                right -= 1
            else:
                newDeck.append(leftArr[0])
                leftArr.pop(0)
                left -= 1

        deck = newDeck
    return deck

def example():
    print(card_shuffle(10))

if __name__ == "__main__":
    example()

File: card-shuffling/card-shuffling.js
const total = 52

function comb(n, k) {
  if (k > n) return 0;
  if (k === 0 || k === n) return 1;
  let res = 1;
  for (let i = 1; i <= k; i++) {
    res *= (n - i + 1) / i;
  }
  return res;
}

const probs = {}
for (let i = 0; i<total+1; i++){
    const prob = comb(total, i) * 0.5**total
    probs[i] = prob
}

const suits = ["H", "D", "C", "S"]
const ranks = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]


var deck = suits.flatMap(suit =>
  ranks.map(rank => `${rank} ${suit}`)
);

function weightedChoice(items, weights) {
  const total = weights.reduce((a, b) => a + b, 0);
  let r = Math.random() * total;

  for (let i = 0; i < items.length; i++) {
    if (r < weights[i]) return items[i];
    r -= weights[i];
  }
}

function split_in_half(){
    const num = Number(weightedChoice(Object.keys(probs), Object.values(probs)))
    return num
}

function card_shuffle(n){
    for (let i = 0; i < n; i++){
        let left = split_in_half()
        let right = total - left
        const leftArr = deck.slice(0, left)
        const rightArr = deck.slice(left)

        const newDeck = []
        for (let j = 0; j < total; j++){
            const takeLeft = Math.random() < left / (left + right);
            if (takeLeft) {
                newDeck.push(leftArr.shift());
                left--;
            } else {
                newDeck.push(rightArr.shift());
                right--;
            }
        }
        deck = newDeck
    }
    return deck
}

console.log(card_shuffle(10))